src/3rd_party.c

```diff
@@ -0,0 +1,173 @@
+/*
+ *   Software Updater - client side
+ *
+ *      Copyright © 2019 Intel Corporation.
+ *
+ *   This program is free software: you can redistribute it and/or modify
+ *   it under the terms of the GNU General Public License as published by
+ *   the Free Software Foundation, version 2 or later of the License.
+ *
+ *   This program is distributed in the hope that it will be useful,
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *   GNU General Public License for more details.
+ *
+ *   You should have received a copy of the GNU General Public License
+ *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#define _GNU_SOURCE
+#include "3rd_party.h"
+#include "config.h"
+#include "lib/log.h"
+#include "swupd.h"
+
+static char *cmdline_option_repo = NULL;
+
+#define CHOOSE_ADD 9000
+#define CHOOSE_LIST 9001
+#define CHOOSE_REMOVE 9003
+static bool cmdline_option_add;
+static bool cmdline_option_list;
+static bool cmdline_option_remove;
+
+#ifdef THIRDPARTY
+
+void get_repo_path(char **repo_path, char *repo_name)
+{
+	string_or_die(repo_path, "%s%s", THIRDPARTY_REPO_PREFIX, repo_name);
+}
+
+static void print_help(void)
+{
+	print("Usage:\n");
+	print("   swupd 3rd-party [OPTION...]\n\n");
+
+	global_print_help();
+
+	print("Options:\n");
+	print("   --add <repo_name>		Add third party repo\n");
+	print("   --list			List third party repo(s)");
+	print("   --remove <repo_name>		Remove third party repo\n");
+	print("\n");
+}
+
+static const struct option prog_opts[] = {
+	{ "repo", required_argument, NULL, 'R' },
+	{ "add", no_argument, 0, CHOOSE_ADD },
+	{ "list", no_argument, 0, CHOOSE_LIST },
+	{ "remove", no_argument, 0, CHOOSE_REMOVE },
+};
+
+static bool parse_opt(int opt, UNUSED_PARAM char *optarg)
+{
+	switch (opt) {
+	case 'R':
+		if (!optarg) {
+			error("Invalid repo: %s\n\n", optarg);
+		}
+		cmdline_option_repo = optarg;
+		return true;
+	case CHOOSE_ADD:
+		cmdline_option_add = optarg_to_bool(optarg);
+		return true;
+	case CHOOSE_REMOVE:
+		cmdline_option_remove = optarg_to_bool(optarg);
+		return true;
+	case CHOOSE_LIST:
+		cmdline_option_list = optarg_to_bool(optarg);
+		return true;
+	default:
+		return false;
+	}
+	return false;
+}
+
+static const struct global_options opts = {
+	prog_opts,
+	sizeof(prog_opts) / sizeof(struct option),
+	parse_opt,
+	print_help,
+};
+
+static bool parse_options(int argc, char **argv)
+{
+	int ind = global_parse_options(argc, argv, &opts);
+	int opt_mask = 0;
+
+	if (ind < 0) {
+		return false;
+	}
+
+	if (!cmdline_option_repo) {
+		if (cmdline_option_add || cmdline_option_remove) {
+			error("The --repo option is a required argument\n\n");
```

Comment:
same for --repo. I prefer:

`swupd 3rd-party add repo_name url`

Reply:
Actually I was thinking this should be an option for sure. Usually command are actions.

It works perfectly in our case as then we can make it required for remove, add and not have it for list.

Let me know what you feel